packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,186 @@
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        Promise.resolve().then(() => {
+          if (wasFulfilled) {
+            if (watcher.onFulfilled) {
+              watcher.onFulfilled(value, ...args);
+            }
+          } else {
+            if (watcher.onRejected) {
+              watcher.onRejected(value, ...args);
+            } else {
+              throw value;
+            }
+          }
+        });
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises');
+    let deadPromises;
+    if (deadPromisesRaw) {
+      deadPromises = new Set(deadPromisesRaw.split(','));
+      syscall.vatstoreDelete('deadPromises');
+    } else {
+      deadPromises = new Set();
+    }
+    for (const [vpid, watches] of watchedPromiseTable.entries()) {
+      if (deadPromises.has(vpid)) {
+        watchedPromiseTable.delete(vpid);
+        for (const watch of watches) {
+          const [watcher, ...args] = watch;
+          if (watcher.onRejected) {
+            Promise.resolve().then(() => {
+              watcher.onRejected('vat upgraded', ...args);
+            });
+          }
+        }
+      } else {
+        const p = convertSlotToVal(vpid);
+        pseudoThen(p, vpid);
+      }
+    }
+  }
+
+  // defined promise watcher objects indexed by kindHandle
+  let promiseWatcherByKindTable;
+
+  function preparePromiseWatcherTables() {
+    let watcherTableID = syscall.vatstoreGet('watcherTableID');
+    if (watcherTableID) {
+      promiseWatcherByKindTable = convertSlotToVal(watcherTableID);
+    } else {
+      promiseWatcherByKindTable = makeScalarBigMapStore(
+        'promiseWatcherByKind',
+        { durable: true },
+      );
+      watcherTableID = convertValToSlot(promiseWatcherByKindTable);
+      syscall.vatstoreSet('watcherTableID', watcherTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watcherTableID);
+    }
+
+    let watchedPromiseTableID = syscall.vatstoreGet('watchedPromiseTableID');
+    if (watchedPromiseTableID) {
+      watchedPromiseTable = convertSlotToVal(watchedPromiseTableID);
+    } else {
+      watchedPromiseTable = makeScalarBigMapStore('watchedPromises', {
+        keySchema: M.string(),
```

Comment:
let's add a `// vpid` comment to the schema line, a cheap way to help orient future maintainers